Mnist softmax is learning

jburroni opened the pull request:
I think we can merge this

jburroni committed:
use SoftMax. merge of
e29e40e7c5c18a8a7246658885df0a403818f114
ee57fb1d9b8b792a8185721c96c7a269d254c266
0f90d938032d32053b16ff848db6684bbb2bcefa

jburroni committed:
Merge branch 'master' into mnist-learning

# Conflicts:
#	Packages/TensorFlow-Kernel.pck.st

jburroni committed:
allows to use other output than 0 when composing operations

jburroni committed:
learn using the singleLayer softmax

jburroni committed:
test useOutput:

jburroni committed:
expectedLabel was defined twice

jburroni committed:
testLearning. It’s learning :)

jburroni committed:
new operation: findMaxOn:

jburroni committed:
use finMaxOn: after prediction

jburroni committed:
prediction should be Int64 and not Int32. This is kind of dangerous as no error is raised

gerasdf commented on Packages/TensorFlow-Examples.pck.st:
```diff
@@ -804,25 +836,27 @@ initialize
 		hidden1 _ (image * weights1 + biases1) rectified.
 		hidden2 _ (hidden1 * weights2 + biases2) rectified.
-		hidden2 * weights3 + biases3].
+		netInput _ hidden2 * weights3 + biases3.
```
why is it called netInput? it's roughly at then of the forward path, I'd intuitively say it's closer to being the output of the net, than an input...

jburroni commented:
Because it's the net input for the activation function. That is the typical
name given in the literature. Logits makes sense only for sigmoid
activation
On Sat, Feb 11, 2017 at 18:11 gerasdf <notifications@github.com> wrote:

> *@gerasdf* commented on this pull request.
>
>
>
>
> ------------------------------
>
>
>
>
> In Packages/TensorFlow-Examples.pck.st
> <https://github.com/Cuis-Smalltalk/Machine-Learning/pull/12#pullrequestreview-21396058>
> :
>
>
> >  		input _ image.
>
>  		hidden1 _ (image * weights1 + biases1) rectified.
>
>  		hidden2 _ (hidden1 * weights2 + biases2) rectified.
>
> -		hidden2 * weights3 + biases3].
>
> +		netInput _ hidden2 * weights3 + biases3.
>
>
>
> why is it called netInput? it's roughly at then of the forward path, I'd
> intuitively say it's closer to being the output of the net, than an input...
>
>
>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Cuis-Smalltalk/Machine-Learning/pull/12#pullrequestreview-21396058>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AChv6h9TdFqawlX59g-d1A30AjDzSBm8ks5rbkA5gaJpZM4L-T_2>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

jburroni commented:
Following with the biological model, it's the input to the neuron
On Sat, Feb 11, 2017 at 18:26 Javier Burroni <javier.burroni@gmail.com>
wrote:

> Because it's the net input for the activation function. That is the
> typical name given in the literature. Logits makes sense only for sigmoid
> activation
>
> On Sat, Feb 11, 2017 at 18:11 gerasdf <notifications@github.com> wrote:
>
> *@gerasdf* commented on this pull request.
>
>
>
>
> ------------------------------
>
>
>
>
> In Packages/TensorFlow-Examples.pck.st
> <https://github.com/Cuis-Smalltalk/Machine-Learning/pull/12#pullrequestreview-21396058>
> :
>
>
> >  		input _ image.
>
>  		hidden1 _ (image * weights1 + biases1) rectified.
>
>  		hidden2 _ (hidden1 * weights2 + biases2) rectified.
>
> -		hidden2 * weights3 + biases3].
>
> +		netInput _ hidden2 * weights3 + biases3.
>
>
>
> why is it called netInput? it's roughly at then of the forward path, I'd
> intuitively say it's closer to being the output of the net, than an input...
>
>
>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Cuis-Smalltalk/Machine-Learning/pull/12#pullrequestreview-21396058>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AChv6h9TdFqawlX59g-d1A30AjDzSBm8ks5rbkA5gaJpZM4L-T_2>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

gerasdf committed:
Merge branch 'master' into mnist-learning

gerasdf committed:
I guess the test now makes more sence, the comment no longer applies... I hope :-)

gerasdf committed:
The conditiona was not tested, I guess it was just a distraction

gerasdf committed:
Just prettier

gerasdf committed:
using #output instead of #output:

gerasdf committed:
just formatting

gerasdf commented:
That's what I thoght... I still don't think the literature is using a good
name... Lets keep it, but lets also think about it.

I don't think netInput is a good name because:

Net may also refer to the net (the neural net).
Input makes the reader think that it's an... input (to the system).

I understand this values are the net (pure) input to the following
operation (the activation). Maybe activationInput? unfoldedOutput?
unclassifiedOutput?

I also see that netInput is very deep in the literature... so... maybe...
oh well...

cheers!

On Sat, Feb 11, 2017 at 8:26 PM, jburroni <notifications@github.com> wrote:

> Because it's the net input for the activation function. That is the typical
> name given in the literature. Logits makes sense only for sigmoid
> activation
> On Sat, Feb 11, 2017 at 18:11 gerasdf <notifications@github.com> wrote:
>
> > *@gerasdf* commented on this pull request.
> >
> >
> >
> >
> > ------------------------------
> >
> >
> >
> >
> > In Packages/TensorFlow-Examples.pck.st
> > <https://github.com/Cuis-Smalltalk/Machine-Learning/
> pull/12#pullrequestreview-21396058>
> > :
> >
> >
> > > input _ image.
> >
> > hidden1 _ (image * weights1 + biases1) rectified.
> >
> > hidden2 _ (hidden1 * weights2 + biases2) rectified.
> >
> > - hidden2 * weights3 + biases3].
> >
> > + netInput _ hidden2 * weights3 + biases3.
> >
> >
> >
> > why is it called netInput? it's roughly at then of the forward path, I'd
> > intuitively say it's closer to being the output of the net, than an
> input...
> >
> >
> >
> >
> > —
> > You are receiving this because you were assigned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Cuis-Smalltalk/Machine-Learning/
> pull/12#pullrequestreview-21396058>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AChv6h9TdFqawlX59g-
> d1A30AjDzSBm8ks5rbkA5gaJpZM4L-T_2>
> > .
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Cuis-Smalltalk/Machine-Learning/pull/12#issuecomment-279183535>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAfL7oz6jkNQjG0L_OVJ3GhCjy3IpfqYks5rbkOugaJpZM4L-T_2>
> .
>

gerasdf committed:
Using named sub-graphs for easier readability

gerasdf committed:
Unified use of #sparseSoftmaxCrossEntropyWithLogits: in loss and learning graphs.

This also means there's a single input for expectedLabel.
Also added a named subgraph for loss and learning parts, for nicer display

gerasdf committed:
Named subgraph for inference path. Just for nicer display in a notebook

gerasdf committed:
forgot to commit this after unification of expectedLabd and expectedForComparision

gerasdf committed:
using #output instead of #output:

gerasdf committed:
Factored out some code, AND subgraphs are not created on -p. This doesn't work too well with our graphs

gerasdf committed:
Moved MNIST file tests (slow) from TensorFlowCAPITestSlowTest to TensorFlowExamplesSlowTests

gerasdf committed:
Renamed TensorFlowCAPITestSlowTests to TensorFlowCAPISlowTests

gerasdf commented on Packages/TensorFlow-Examples.pck.st:
```diff
@@ -1090,30 +1090,30 @@ initializeInferenceGraph
-			expectedForComparison _ expected.
-			(netInput sparseSoftmaxCrossEntropyWithLogits: expected) meanOn: #(0) asInt32Tensor]
-		inputTypes: {TF_Tensor typeInt32}! !
+			expectedLabel _ expected.
```
Javi, check how I unified part of the loss calculation path and the learning path.
Both now use the input (expectedLabel) and same activation node, the loss side uses output #0 from the softmax and the learning side uses output #1.

I believe it's still correct, but a bit simpler and a bit more aligned with the ideas of the softmax node, which has two outputs.

If you confirm it's good and like it, merge it :-)

:+1:
